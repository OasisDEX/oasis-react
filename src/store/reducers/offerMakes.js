import { createAction, handleActions } from "redux-actions";
import { fromJS } from "immutable";
import { createPromiseActions } from "../../utils/createPromiseActions";
import offerMakes from "../selectors/offerMakes";
import { change, formValueSelector, initialize } from "redux-form/immutable";
import { TX_OFFER_MAKE } from "./transactions";
import web3 from "../../bootstrap/web3";
import balances from "../selectors/balances";
import { fulfilled, pending, rejected } from "../../utils/store";
import offerMakeToFormName from "../../utils/offers/offerMakeToFormName";

import throttle from "lodash/throttle";

import { defer } from "../deferredThunk";

import { MAKE_BUY_OFFER, MAKE_SELL_OFFER } from "constants";
import { handleTransaction } from "../../utils/transactions/handleTransaction";
import {
  getMarketContractInstance,
  getMarketNoProxyContractInstance
} from "../../bootstrap/contracts";
import { DEFAULT_GAS_LIMIT, DEFAULT_GAS_PRICE } from "../../constants";
import transactions from "../selectors/transactions";
import isNumericAndGreaterThanZero from "../../utils/numbers/isNumericAndGreaterThanZero";
import getTokenByAddress from "../../utils/tokens/getTokenByAddress";
import {
  convertToTokenPrecision
} from "../../utils/conversion";

const initialState = fromJS({
  makeBuyOffer: {
    type: MAKE_BUY_OFFER,
    isOfferMakeModalOpen: false,
    transactionGasCostEstimate: null
  },
  makeSellOffer: {
    type: MAKE_SELL_OFFER,
    isOfferMakeModalOpen: false,
    transactionGasCostEstimate: null
  },
  activeOfferMakeType: null
});

const buyMaxEpic = (
  offerMakeType,
  { activeQuoteTokenBalance = balances.activeQuoteTokenBalance } = {}
) => (dispatch, getState) => {
  const balance = activeQuoteTokenBalance(getState());
  const total = web3.fromWei(balance);
  const totalSerialized = web3
    .toBigNumber(total)
    .toFixed()
    .toString();
  dispatch(localChange(offerMakeType, "total", totalSerialized));
  dispatch(defer(totalFieldValueChangedEpic, offerMakeType, totalSerialized));
};

const sellMaxEpic = (
  offerMakeType,
  {
    activeBaseTokenBalance = balances.activeBaseTokenBalance,
    currentFormValues = offerMakes.currentFormValues
  } = {}
) => (dispatch, getState) => {
  const balance = activeBaseTokenBalance(getState());
  const formName = offerMakeToFormName(offerMakeType);

  const { price } = currentFormValues(getState())(formName);

  const usersBaseTokenBalanceBN = web3.toBigNumber(web3.fromWei(balance));
  const priceBN = web3.toBigNumber(price.toString());

  const totalSerialized = usersBaseTokenBalanceBN
    .mul(priceBN)
    .toFixed()
    .toString();
  dispatch(change(formName, "total", totalSerialized));
  dispatch(defer(totalFieldValueChangedEpic, offerMakeType, totalSerialized));
};

/**
 *  Create new offer transaction actions
 */

//TODO: window? why not direct call?
const makeOfferTransaction = createAction(
  "OFFER_MAKES/MAKE_OFFER_TRANSACTION",
  ({
    payAmount,
    payToken,
    buyAmount,
    buyToken,
    isCloseMatchingEnabled,
    gasPrice = DEFAULT_GAS_PRICE
  }) => {

    const offer = getMarketNoProxyContractInstance().offer['uint256,address,uint256,address,uint256,bool'];

    return new Promise((resolve, reject) =>
      offer(
        payAmount,
        payToken,
        buyAmount,
        buyToken,
        0,
        isCloseMatchingEnabled,
        {
          gasPrice
        },
        (err, res) =>
      {
        if (err) {
          reject(err);
        } else {
          resolve(res);
        }
      })
    );
  }
);

const makeOffer = createPromiseActions("OFFER_MAKES/MAKE_OFFER");

//TODO: To many duties at once. Offer form and offer making logic needlessly coupled here...
const makeOfferEpic = (offerMakeType, withCallbacks = {}) => async (
  dispatch,
  getState
) => {
  dispatch(makeOffer.pending());

  const activeOfferMake = offerMakes.activeOfferMakePure(
    getState(),
    offerMakeToFormName(offerMakeType)
  );

  const makeOfferPayload = {
    payAmount: convertToTokenPrecision(
      activeOfferMake.getIn(["offerData", "payAmount"]),
      activeOfferMake.get("sellToken")
    ),
    buyAmount: convertToTokenPrecision(
      activeOfferMake.getIn(["offerData", "buyAmount"]),
      activeOfferMake.get("buyToken")
    ),
    payToken: activeOfferMake.get("sellTokenAddress"),
    buyToken: activeOfferMake.get("buyTokenAddress"),
    isCloseMatchingEnabled: true
  };

  return handleTransaction({
    dispatch,
    transactionDispatcher: () =>
      dispatch(makeOfferTransaction(makeOfferPayload)),
    transactionType: TX_OFFER_MAKE,
    txMeta: { offerMakeType, makeOfferPayload },
    withCallbacks
  });
};

/**
 * Initial setup actions
 */

const setActiveOfferMakeType = createAction(
  "OFFER_MAKES/SET_ACTIVE_OFFER_MAKE_TYPE",
  makeType => makeType
);
const resetActiveOfferMakeType = createAction(
  "OFFER_MAKES/RESET_ACTIVE_OFFER_MAKE_TYPE"
);

const initializeOfferMakeFormEpic = offerMakeType => dispatch => {
  const formName = offerMakeToFormName(offerMakeType);
  dispatch(
    initialize(formName, {
      price: null,
      volume: null,
      total: null
    })
  );
};

/**
 *  Set offer make modal  open / closed
 */

const setOfferMakeModalOpen = createAction(
  "OFFER_MAKES/SET_MODAL_OPEN",
  offerMakeType => ({ formName: offerMakeToFormName(offerMakeType) })
);

const setOfferMakeModalOpenEpic = offerMakeType => dispatch => {
  dispatch(setActiveOfferMakeType(offerMakeType));
  dispatch(updateTransactionGasCostEstimateEpicThrottled(offerMakeType));
  dispatch(setOfferMakeModalOpen(offerMakeType));
};

const setOfferMakeModalClosed = createAction(
  "OFFER_MAKES/SET_MODAL_CLOSED",
  offerMakeType => ({ formName: offerMakeToFormName(offerMakeType) })
);

const setOfferMakeModalClosedEpic = offerMakeType => dispatch => {
  dispatch(setOfferMakeModalClosed(offerMakeType));
  dispatch(resetActiveOfferMakeType());
  dispatch(resetActiveOfferMakeGasCostEstimate());
};

const localFormValueSelector = offerMakeType =>
  formValueSelector(offerMakeToFormName(offerMakeType));

const localChange = (offerMakeType, ...args) =>
  change(offerMakeToFormName(offerMakeType), ...args);

/***
 * Form fields changed actions
 */
const priceFieldChangedEpic = (
  offerMakeType,
  value,
  { formValueSelector = localFormValueSelector } = {}
) => (dispatch, getState) => {
  const volume = formValueSelector(offerMakeType)(getState(), "volume");
  if (
    !value ||
    (isNaN(value) || parseFloat(value) === 0) ||
    (isNaN(volume) || parseFloat(volume) === 0)
  ) {
    dispatch(localChange(offerMakeType, "total", null));
  } else {
    dispatch(
      localChange(
        offerMakeType,
        "total",
        web3
          .toBigNumber(volume)
          .mul(value)
          .toFixed()
          .toString()
      )
    );
  }
  dispatch(defer(updateTransactionGasCostEstimateEpicThrottled, offerMakeType));
};

const volumeFieldValueChangedEpic = (
  offerMakeType,
  value,
  { formValueSelector = localFormValueSelector } = {}
) => (dispatch, getState) => {
  const price = formValueSelector(offerMakeType)(getState(), "price");

  if (
    isNaN(value) ||
    parseFloat(value) <= 0 ||
    (isNaN(price) || parseFloat(price) === 0)
  ) {
    dispatch(localChange(offerMakeType, "total", "0"));
  } else {
    // dispatch(localChange(offerMakeType, 'volume', value));
    dispatch(
      localChange(
        offerMakeType,
        "total",
        web3
          .toBigNumber(value)
          .mul(price)
          .toFixed()
          .toString()
      )
    );
  }
  dispatch(defer(updateTransactionGasCostEstimateEpicThrottled, offerMakeType));
};

const totalFieldValueChangedEpic = (
  offerMakeType,
  value,
  { formValueSelector = localFormValueSelector } = {}
) => (dispatch, getState) => {
  const price = formValueSelector(offerMakeType)(getState(), "price");
  if (
    !isNaN(value) &&
    parseFloat(value) > 0 &&
    (!isNaN(price) && parseFloat(price) > 0)
  ) {
    dispatch(
      localChange(
        offerMakeType,
        "volume",
        web3
          .toBigNumber(value)
          .div(price)
          .toFixed()
          .toString()
      )
    );
    dispatch(
      defer(updateTransactionGasCostEstimateEpicThrottled, offerMakeType)
    );
  } else {
    dispatch(localChange(offerMakeType, "volume", "0"));
  }
};

/**
 *  New offer make gas estimate
 */

const resetActiveOfferMakeGasCostEstimate = createAction(
  "OFFER_MAKES/RESET_TRANSACTION_GAS_ESTIMATE"
);

const getTransactionGasEstimate = createAction(
  "OFFER_MAKES/GET_TRANSACTION_GAS_ESTIMATE",
  (payAmount, payToken, buyAmount, buyToken, toAddress) =>
    new Promise((resolve, reject) => {
      getMarketNoProxyContractInstance().offer['uint256,address,uint256,address,uint256,bool'].estimateGas(
        convertToTokenPrecision(payAmount, getTokenByAddress(payToken)),
        payToken,
        convertToTokenPrecision(buyAmount, getTokenByAddress(buyToken)),
        buyToken,
        0,
        true, // isCloseMatchingEnabled
        { to: toAddress },
        (e, estimation) => {
          if (e) {
            reject({
              payAmount: convertToTokenPrecision(
                payAmount,
                getTokenByAddress(payToken)
              ),
              payToken,
              buyAmount: convertToTokenPrecision(
                buyAmount,
                getTokenByAddress(buyToken)
              ),
              buyToken,
              toAddress
            });
          } else {
            resolve(estimation);
          }
        }
      );
    })
);

const setGasExceedsTheLimitEnabled = createAction(
  "OFFER_MAKES/GAS_EXCEEDS_THE_LIMIT_ENABLED"
);
const setGasExceedsTheLimitDisabled = createAction(
  "OFFER_MAKES/GAS_EXCEEDS_THE_LIMIT_DISABLED"
);

const updateTransactionGasCostEstimateEpic = (
  offerMakeType,
  {
    canMakeOffer = offerMakes.canMakeOffer,
    activeOfferMakePure = offerMakes.activeOfferMakePure
  } = {}
) => async (dispatch, getState) => {
  if (!canMakeOffer(getState(), offerMakeType)) {
    dispatch(resetActiveOfferMakeGasCostEstimate());
    return;
  }

  const offerMake = activeOfferMakePure(
    getState(),
    offerMakeToFormName(offerMakeType)
  );

  const payAmount = offerMake.getIn(["offerData", "payAmount"]),
    payToken = offerMake.get("sellTokenAddress"),
    buyAmount = offerMake.getIn(["offerData", "buyAmount"]),
    buyToken = offerMake.get("buyTokenAddress"),
    toAddress = getMarketContractInstance().address;
  if (
    isNumericAndGreaterThanZero(payAmount) &&
    isNumericAndGreaterThanZero(buyAmount)
  ) {
    const transactionGasCostEstimate = (await dispatch(
      defer(
        getTransactionGasEstimate,
        payAmount,
        payToken,
        buyAmount,
        buyToken,
        toAddress
      )
    )).value;

    const gasLimitInWeiBN = web3.toBigNumber(
      transactions.defaultGasLimit(getState())
    );
    if (gasLimitInWeiBN.lt(transactionGasCostEstimate)) {
      dispatch(setGasExceedsTheLimitEnabled());
    } else {
      dispatch(setGasExceedsTheLimitDisabled());
    }
    return transactionGasCostEstimate;
  }
};

const updateTransactionGasCostEstimateEpicThrottled = throttle(
  updateTransactionGasCostEstimateEpic,
  500
);

const actions = {
  makeOfferEpic,
  setOfferMakeModalClosedEpic,
  setOfferMakeModalOpenEpic,
  priceFieldChangedEpic,
  volumeFieldValueChangedEpic,
  totalFieldValueChangedEpic,
  buyMaxEpic,
  sellMaxEpic,
  initializeOfferMakeFormEpic,
  updateTransactionGasCostEstimateEpic //exported just for testing sake
};

const reducer = handleActions(
  {
    [setOfferMakeModalOpen]: (state, { payload: { formName } }) =>
      state.setIn([formName, "isOfferMakeModalOpen"], true),
    [setOfferMakeModalClosed]: (state, { payload: { formName } }) =>
      state.setIn([formName, "isOfferMakeModalOpen"], false),
    [setActiveOfferMakeType]: (state, { payload }) =>
      state.set("activeOfferMakeType", payload),
    [resetActiveOfferMakeType]: state => state.set("activeOfferMakeType", null),
    [pending(getTransactionGasEstimate)]: state =>
      state
        .set("transactionGasCostEstimateError", false)
        .set("transactionGasCostEstimatePending", true),
    [fulfilled(getTransactionGasEstimate)]: (state, { payload }) =>
      state
        .set("transactionGasCostEstimateError", false)
        .set("transactionGasCostEstimate", payload.toString())
        .set("transactionGasCostEstimatePending", false),
    [rejected(getTransactionGasEstimate)]: state =>
      state
        .set("transactionGasCostEstimateError", true)
        .set("transactionGasCostEstimatePending", false),
    [resetActiveOfferMakeGasCostEstimate]: state =>
      state
        .set("transactionGasCostEstimate", null)
        .set("transactionGasCostEstimateError", null)
        .set("transactionGasCostEstimatePending", null),
    [setGasExceedsTheLimitEnabled]: state =>
      state.set("exceededGasLimit", true),
    [setGasExceedsTheLimitDisabled]: state =>
      state.set("exceededGasLimit", false)
  },
  initialState
);

export default {
  actions,
  reducer
};
