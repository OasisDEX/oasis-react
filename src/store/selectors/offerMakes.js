import { createSelector, createStructuredSelector } from "reselect";
import { formValueSelector } from "redux-form/immutable";
import balances from "./balances";
import transactions from "./transactions";
import markets from "./markets";
import web3 from "../../bootstrap/web3";
import {
  MAKE_BUY_OFFER,
  MAKE_BUY_OFFER_FORM_NAME,
  MAKE_SELL_OFFER,
  MAKE_SELL_OFFER_FORM_NAME
} from "../../constants";
import reselect from "../../utils/reselect";
import {
  offerMakeToFormName,
  formNameToOfferMake
} from '../../utils/offers/offerMakeToFormName'
import tokens from "./tokens";
import network from "./network";
import { fromJS, Map } from "immutable";
import limits from "./limits";
import isNumeric from "../../utils/numbers/isNumeric";
import {memoize} from 'lodash';
import { offerMakeTypeToOfferListName } from '../../utils/offers/offerMakeTypeToOfferListName'
import { getOfferPrice } from '../../utils/offers/getOfferPrice'

const offerMakes = state => state.get("offerMakes");

const formFields = createStructuredSelector({
  [MAKE_BUY_OFFER_FORM_NAME]: reselect.formFieldsSelector(MAKE_BUY_OFFER_FORM_NAME, 'volume', 'price', 'total'),
  [MAKE_SELL_OFFER_FORM_NAME]: reselect.formFieldsSelector(MAKE_SELL_OFFER_FORM_NAME, 'volume', 'price', 'total'),
})

const currentFormValues = createSelector(
  formFields,
  formValues => memoize(formName => formValues[formName])
);

//TODO: move to utils
const toWeiString = amount =>
  web3
  .toBigNumber(web3.toWei(amount))
  .ceil()
  .toString();

const selector = createSelector(
  state => state.a,
  state => state.b,
  (a, b) => ({
    c: a * 2,
    d: b * 3
  })
);

const hasExceededGasLimit = createSelector(
  offerMakes,
  s => Boolean(s.get('exceededGasLimit'))
);


const activeTradingPairOffersData = createSelector(
  rootState => rootState.get("offers"),
  tokens.activeTradingPair,
  (state, activeTradingPair) => state.getIn(["offers", Map(activeTradingPair)])
);

const getNewOfferRankIndex = createSelector(
  (...args) => args[1],
  currentFormValues,
  activeTradingPairOffersData,
  (
    offerFormName,
    currentFormValues,
    currentTradingPairOffersData = fromJS({ buyOffers: [], sellOffers: [] })
  ) => {
    const { price } = currentFormValues(offerFormName);
    if (!parseFloat(String(price))) {
      return;
    }

    const priceBN = web3.toBigNumber(price);
    const formMakeType = formNameToOfferMake(offerFormName);
    const offersSortedByPriceAsc = currentTradingPairOffersData
    .get(offerMakeTypeToOfferListName(formMakeType))
    .map(offer => ({
      price: getOfferPrice(offer, formMakeType),
      offerId: offer.id
    }))
    .sort(
      ({ price: priceA }, { price: priceB }) => (priceA.gte(priceB) ? 1 : -1)
    )
    .map((d, idx) => ({ ...d, idx }));

    for (const { price, offerId, idx } of offersSortedByPriceAsc) {
      switch (priceBN.comparedTo(price)) {
        case -1:
          // console.log('offerId', offerId.toString());
          return offerId;
        case 0: {
          if (idx > 0) {
            const { offerId } = offersSortedByPriceAsc.get(idx - 1);
            // console.log('offerId', offerId.toString());
            return offerId;
          } else {
            const { offerId } = offersSortedByPriceAsc.first();
            // console.log('offerId', offerId.toString());
            return offerId;
          }
        }
      }
    }

    if (offersSortedByPriceAsc.count()) {
      // console.log('last', offersSortedByPriceAsc.last().offerId.toString());
      return offersSortedByPriceAsc.last().offerId;
    } else {
      // console.log('last', 0);
      return 0;
    }
  }
);

const activeOfferMakePure = createSelector(
  (...args) => args[1], //provides original selector argument
  tokens.activeTradingPair,
  network.tokenAddresses,
  currentFormValues,
  (rootState, offerMakeFormName) =>
    getNewOfferRankIndex(rootState, offerMakeFormName),
  (
    offerMakeFormName,
    activeTradingPair,
    tokenAddresses,
    currentFormValues,
    newOfferRankIndex
  ) => {
    const { total, volume } = currentFormValues(offerMakeFormName);
    const { baseToken, quoteToken } = activeTradingPair.toJS
      ? activeTradingPair.toJS()
      : activeTradingPair;
    const offerMakeType = formNameToOfferMake(offerMakeFormName);
    console.assert(
      offerMakeType,
      `Wrong offerMakeFormName: ${offerMakeFormName}`
    );
    const buyToken = offerMakeType === MAKE_BUY_OFFER ? baseToken : quoteToken;
    const sellToken = offerMakeType === MAKE_BUY_OFFER ? quoteToken : baseToken;

    return fromJS({
      baseToken,
      quoteToken,
      buyToken,
      sellToken,
      newOfferRankIndex,
      sellTokenAddress: tokenAddresses.get(sellToken),
      buyTokenAddress: tokenAddresses.get(buyToken),
      offerData: {
        payAmount: toWeiString(
          offerMakeType === MAKE_BUY_OFFER ? total : volume
        ),
        buyAmount: toWeiString(
          offerMakeType === MAKE_BUY_OFFER ? volume : total
        )
      }
    });
  }
);


// const activeOfferMake = activeOfferMakePure;

const activeOfferMakeType = createSelector(offerMakes, state =>
  state.get("activeOfferMakeType")
);

const activeOfferMakeOfferData = createSelector(activeOfferMakePure, state =>
  state.getIn("offerData")
);

const activeOfferMakeOfferOwner = createSelector(offerMakes, state =>
  state.getIn(["activeOfferMake", "offerData", "owner"])
);

const activeOfferMakeBuyToken = createSelector(activeOfferMakePure, state =>
  state.get("buyToken")
);

const activeOfferMakeSellToken = createSelector(activeOfferMakePure, state =>
  state.get("sellToken")
);

const isOfferMakeModalOpen = createSelector(
  offerMakes,
  reselect.getProps,
  (state) => memoize(offerMakeType =>
    state.getIn([
      offerMakeToFormName(offerMakeType),
      "isOfferMakeModalOpen"
    ])
  )
);

const makeFormValuesSelector = formName => {
  return formValueSelector(formName);
};

const hasSufficientTokenAmount = createSelector(
  balances.tokenBalances,
  currentFormValues,
  tokens.activeTradingPair,
  (
    tokenBalances,
    currentFormValues,
    { baseToken, quoteToken }
  ) => memoize(activeOfferMakeType => {
    const { total, volume } = currentFormValues(offerMakeToFormName(activeOfferMakeType));
    if (!volume) {
      return true;
    } else {
      switch (activeOfferMakeType) {
        case MAKE_BUY_OFFER: {
          const usersSellTokenBalanceBN = web3.toBigNumber(
            tokenBalances.get(quoteToken)
          );
          return usersSellTokenBalanceBN.gte(web3.toWei(total));
        }
        case MAKE_SELL_OFFER: {
          const usersSellTokenBalanceBN = web3.toBigNumber(
            tokenBalances.get(baseToken)
          );
          return usersSellTokenBalanceBN.gte(web3.toWei(volume));
        }
      }
    }
  })
);

const isVolumeOrPriceEmptyOrZero = createSelector(
  currentFormValues,
  (currentFormValues) => memoize(offerMakeType => {
    const { volume, price } = currentFormValues(offerMakeToFormName(offerMakeType));
    return !isNumeric(price) ||
      web3.toBigNumber(price).lte(0) ||
      (!isNumeric(volume) || web3.toBigNumber(volume).lte(0))
  })
);


const tokenMinSellLimitInWeiByOfferType = createSelector(
  tokens.activeTradingPair,
  limits.tokenMinSellLimitInWei,
  (activeTradingPair, tokenMinSellLimitInWei) => memoize(offerMakeType => {
    let tokenName = null;
    const activeTradingPairMap = Map(activeTradingPair);
    if (!activeTradingPairMap.has('baseToken') && activeTradingPairMap.has('quoteToken')) {
      return null
    }
    switch (offerMakeType) {
      case MAKE_BUY_OFFER:
        tokenName = activeTradingPairMap.get("quoteToken");
        break;
      case MAKE_SELL_OFFER:
        tokenName = activeTradingPairMap.get("baseToken");
        break;
      default:
        throw new Error("No offer make type provided!");
    }
    return tokenMinSellLimitInWei(tokenName);
  }));

const isOfferBelowLimit = createSelector(
  tokenMinSellLimitInWeiByOfferType,
  currentFormValues,
  (tokenMinSellLimitInWeiByOfferType, currentFormValues) => memoize(offerMakeType => {
    const { total, volume } = currentFormValues(offerMakeToFormName(offerMakeType));
    const tokenMinSellLimitInWei = tokenMinSellLimitInWeiByOfferType(offerMakeType);
    if (tokenMinSellLimitInWei === null) {
      return null;
    }
    let amount = null;
    switch (offerMakeType) {
      case MAKE_BUY_OFFER:
        amount = total;
        break;
      case MAKE_SELL_OFFER:
        amount = volume;
        break;
      default:
        throw new Error("No offer make type provided!");
    }

    if (
      isNumeric(tokenMinSellLimitInWei) &&
      isNumeric(amount) &&
      parseFloat(amount) > 0
    ) {
      return web3.toBigNumber(web3.fromWei(tokenMinSellLimitInWei)).gt(amount);
    }
  })
);

const isTotalOverTheTokenLimit = createSelector(
  limits.activeTradingPairQuoteTokenMaxLimitInWei,
  (rootState, offerMakeType) =>
    makeFormValuesSelector(offerMakeToFormName(offerMakeType))(
      rootState, "total",
    ),
  (tokenMaxSellLimitInWei, total) => {
    if (tokenMaxSellLimitInWei === null) {
      return null;
    } else if (
      isNumeric(tokenMaxSellLimitInWei) && isNumeric(total) && parseFloat(total) > 0
    ) {
      return web3.toBigNumber(web3.fromWei(tokenMaxSellLimitInWei)).lte(total);
    }
  }
);

const checkTokenEnabled = createSelector(
  s => s,
  (...args) => args[1],
  (...args) => args[2],
  (rootState, offerType, skipTokenEnabledCheck) => {
    return (
      skipTokenEnabledCheck ||
      balances.tokenAllowanceStatusForActiveMarket(rootState, {
        tokenName: activeOfferMakeSellToken(
          rootState,
          offerMakeToFormName(offerType)
        )
      })
    );
  },
);

const canMakeOffer = createSelector(
  hasSufficientTokenAmount,
  rootState => transactions.canSendTransaction(rootState),
  markets.isBuyEnabled,
  checkTokenEnabled,
  isVolumeOrPriceEmptyOrZero,
  isOfferBelowLimit,
  isTotalOverTheTokenLimit,
  (s, offerMakeType) => offerMakeType,
  (
    hasSufficientTokenAmount,
    canSendTransaction,
    isBuyEnabled,
    isTokenEnabled,
    isVolumeOrPriceEmptyOrZero,
    isBelowLimit,
    isOverTheLimit,
    offerMakeType,
  ) => {
    if (
      isVolumeOrPriceEmptyOrZero(offerMakeType) ||
      !hasSufficientTokenAmount(offerMakeType) ||
      // !canSendTransaction ||
      !isBuyEnabled ||
      !isTokenEnabled ||
      isBelowLimit(offerMakeType) ||
      isOverTheLimit
    ) {
      return false;
    } else {
      return hasSufficientTokenAmount(offerMakeType);
    }
  }
);

const isOfferActive = createSelector(offerMakes, state =>
  state.get("isOfferActive")
);

const transactionGasCostEstimate = createSelector(offerMakes, state =>
  state.get("transactionGasCostEstimate")
);

const checkingIfOfferIsActive = createSelector(offerMakes, state =>
  state.get("checkingIfOfferActive")
);

const gasEstimatePending = createSelector(offerMakes, state =>
  state.get("transactionGasCostEstimatePending")
);

const getActiveOfferMakeAllowanceStatus = createSelector(
  (rootState, offerType) => {
    return balances.tokenAllowanceStatusForActiveMarket(rootState, {
      tokenName: activeOfferMakeSellToken(
        rootState,
        offerMakeToFormName(offerType)
      )
    });
  },
  status => Boolean(status)
);

const isMakeBuyOfferPriceSet = createSelector(
  rootState => currentFormValues(rootState)(MAKE_BUY_OFFER_FORM_NAME),
  ({ price }) => Boolean(price)
);
const isMakeSellOfferPriceSet = createSelector(
  rootState => currentFormValues(rootState)(MAKE_SELL_OFFER_FORM_NAME),
  ({ price }) => Boolean(price)
);


export default {
  state: offerMakes,
  selector,
  activeOfferMakePure,
  activeOfferMakeType,
  activeOfferMakeOfferData,
  isOfferMakeModalOpen,
  activeOfferMakeBuyToken,
  activeOfferMakeSellToken,
  activeOfferMakeOfferOwner,
  canMakeOffer,
  hasSufficientTokenAmount,
  transactionGasCostEstimate,
  isVolumeOrPriceEmptyOrZero,
  isOfferActive,
  checkingIfOfferIsActive,
  currentFormValues,
  gasEstimatePending,
  getActiveOfferMakeAllowanceStatus,
  isMakeBuyOfferPriceSet,
  isMakeSellOfferPriceSet,
  isOfferBelowLimit,
  hasExceededGasLimit,
  isTotalOverTheTokenLimit
};
