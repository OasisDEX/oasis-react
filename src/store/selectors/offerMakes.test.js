import offerMakes from "./offerMakes";
import { Map } from "immutable";
import { TOKEN_MAKER, TOKEN_WRAPPED_ETH } from '../../constants';
import { fromJS } from 'immutable';

const state = fromJS({
  tokens: Map({
    activeTradingPair: {
      baseToken: "MKR",
      quoteToken: "W-ETH"
    }
  }),
  network: Map({
    activeNetworkId: 42
  }),
  form: {
    makeBuyOffer: {
      values: {
        price: "1",
        volume: "1",
        total: "1"
      }
    },
    makeSellOffer: {
      values: {
        price: "1",
        volume: "1",
        total: "1"
      }
    }
  },
  offers: {
    [Map({ baseToken: TOKEN_WRAPPED_ETH, quoteToken: TOKEN_MAKER })]: {
      loadingBuyOffers: "SYNC_STATUS_COMPLETED",
      initialSyncMeta: {
        syncStartBlockNumber: 8449045,
        syncEndBlockNumber: 8449045,
        syncTimestamps: { syncStartTimestamp: null, syncEndTimestamp: null }
      },
      sellOfferCount: 0,
      loadingSellOffers: "SYNC_STATUS_COMPLETED",
      sellOffers: [],
      initialSyncStatus: "SYNC_STATUS_COMPLETED",
      buyOfferCount: 2,
      reSyncOffersSet: [],
      buyOffers: [
        {
          id: "785",
          owner: "0xf0e90739550992fcf37fe4dcb0b47708ca0ff609",
          buyWhichTokenAddress: "0xaaf64bfcc32d0f15873a02163e7e500671a4ffcd",
          buyWhichToken: "MKR",
          sellWhichTokenAddress: "0xd0a1e359811322d97991e03f863a0c30c2cf029c",
          sellWhichToken: "W-ETH",
          buyHowMuch: "16400000000000000000",
          sellHowMuch: "16400000000000000000",
          buyHowMuch_filter: 16400000000000000000,
          sellHowMuch_filter: 16400000000000000000,
          ask_price: "1",
          bid_price: "1",
          ask_price_sort: 1,
          bid_price_sort: 1
        },
        {
          id: "1312",
          owner: "0xfce1bae0f16c07377dbb680fcd0aa94b011eb7af",
          buyWhichTokenAddress: "0xaaf64bfcc32d0f15873a02163e7e500671a4ffcd",
          buyWhichToken: "MKR",
          sellWhichTokenAddress: "0xd0a1e359811322d97991e03f863a0c30c2cf029c",
          sellWhichToken: "W-ETH",
          buyHowMuch: "4000000000000000000",
          sellHowMuch: "4000000000000000000",
          buyHowMuch_filter: 4000000000000000000,
          sellHowMuch_filter: 4000000000000000000,
          ask_price: "1",
          bid_price: "1",
          ask_price_sort: 1,
          bid_price_sort: 1
        },
      ]
    }
  }
});

test("activeOfferMakePure/makeBuyOffer", () => {
  expect(
    offerMakes.activeOfferMakePure(state, "makeBuyOffer")
  ).toMatchSnapshot();
});

test("activeOfferMakePure/makeSellOffer", () => {
  expect(
    offerMakes.activeOfferMakePure(state, "makeSellOffer")
  ).toMatchSnapshot();
});

test("activeOfferMakeBuyToken/makeBuyOffer", () => {
  expect(
    offerMakes.activeOfferMakeBuyToken(state, "makeBuyOffer")
  ).toMatchSnapshot();
});

test("activeOfferMakeBuyToken/makeSellOffer", () => {
  expect(
    offerMakes.activeOfferMakeBuyToken(state, "makeSellOffer")
  ).toMatchSnapshot();
});
