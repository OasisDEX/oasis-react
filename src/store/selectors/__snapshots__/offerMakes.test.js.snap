// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activeOfferMakeBuyToken/makeBuyOffer 1`] = `"MKR"`;

exports[`activeOfferMakeBuyToken/makeSellOffer 1`] = `"W-ETH"`;

exports[`activeOfferMakePure/makeBuyOffer 1`] = `
Immutable.Map {
  baseToken: "MKR",
  quoteToken: "W-ETH",
  buyToken: "MKR",
  sellToken: "W-ETH",
  newOfferRankIndex: 0,
  sellTokenAddress: "0xd0a1e359811322d97991e03f863a0c30c2cf029c",
  buyTokenAddress: "0xaaf64bfcc32d0f15873a02163e7e500671a4ffcd",
  offerData: Immutable.Map {
    payAmount: "1000000000000000000",
    buyAmount: "1000000000000000000",
  },
}
`;

exports[`activeOfferMakePure/makeSellOffer 1`] = `
Immutable.Map {
  baseToken: "MKR",
  quoteToken: "W-ETH",
  buyToken: "W-ETH",
  sellToken: "MKR",
  newOfferRankIndex: 0,
  sellTokenAddress: "0xaaf64bfcc32d0f15873a02163e7e500671a4ffcd",
  buyTokenAddress: "0xd0a1e359811322d97991e03f863a0c30c2cf029c",
  offerData: Immutable.Map {
    payAmount: "1000000000000000000",
    buyAmount: "1000000000000000000",
  },
}
`;
